Unable to use DSR mode with Geneve

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When running using DSR mode with Geneve, I encounter some errors in my logs. Initially, my cluster never lost all load balancing ability, it seemed to rollback to SNAT mode. However, now it seems to have lost connectivity and I haven't been able to replicate the SNAT thing again.

### Cilium Version

1.14.0-snapshot.2

### Kernel Version

Linux rn6-k8s.va.internal.rocketnode.net 5.14.15-1.el8.elrepo.x86_64 #1 SMP Tue Oct 26 11:45:20 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.10", GitCommit:"5c1d2d4295f9b4eb12bfbf6429fdf989f2ca8a02", GitTreeState:"clean", BuildDate:"2023-01-18T19:08:10Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

_No response_

### Relevant log output

```shell
level=warning msg="JoinEP: Failed to compile" debug=true error="Failed to compile bpf_host.o: exit status 1" params="&{Source:bpf_host.c Output:bpf_host.o OutputType:obj Options:[]}" subsys=datapath-loader
level=error msg="BPF template object creation failed" bpfHeaderfileHash=8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea error="failed to compile template program /var/run/cilium/state/templates/8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea: Failed to compile bpf_host.o: exit status 1" subsys=datapath-loader
level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=4 endpointID=1441 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea: Failed to compile bpf_host.o: exit status 1" identity=1 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="generating BPF for endpoint failed, keeping stale directory" containerID= datapathPolicyRevision=0 desiredPolicyRevision=4 endpointID=1441 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea: Failed to compile bpf_host.o: exit status 1" file-path=1441_next_fail identity=1 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=warning msg="Regeneration of endpoint failed" bpfCompilation=59.82442ms bpfLoadProg=0s bpfWaitForELF=60.000979ms bpfWriteELF=0s containerID= datapathPolicyRevision=0 desiredPolicyRevision=4 endpointID=1441 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea: Failed to compile bpf_host.o: exit status 1" identity=1 ipv4= ipv6= k8sPodName=/ mapSync="4.26µs" policyCalculation="22.14µs" prepareBuild="467.938µs" proxyConfiguration="8.521µs" proxyPolicyCalculation="2.794µs" proxyWaitForAck=0s reason="one or more identities created or deleted" subsys=endpoint total=60.709101ms waitingForCTClean="2.374µs" waitingForLock="7.473µs"
level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=4 endpointID=1441 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/8e852bc9424a68c770fa41787b003eaeaf1f7e6be6bd7ce7e533c197eaf71aea: Failed to compile bpf_host.o: exit status 1" identity=1 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
level=error msg="Failed to compile bpf_host.o: exit status 1" compiler-pid=2028 linker-pid=2029 subsys=datapath-loader
level=warning msg="In file included from /var/lib/cilium/bpf/bpf_host.c:50:" subsys=datapath-loader
level=warning msg="/var/lib/cilium/bpf/lib/nodeport.h:1518:28: error: implicit conversion loses integer precision: 'unsigned long' to '__u16' (aka 'unsigned short') [-Werror,-Wimplicit-int-conversion]" subsys=datapath-loader
level=warning msg="                sizeof(struct genevehdr) + ETH_HLEN;" subsys=datapath-loader
level=warning msg="                ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~" subsys=datapath-loader
level=warning msg="1 error generated." subsys=datapath-loader
```


### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -820,7 +820,7 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
 		// Use templating for ETH_HLEN only if there is any L2-less device
 		if !mac.HaveMACAddrs(option.Config.GetDevices()) {
 			// L2 hdr len (for L2-less devices it will be replaced with "0")
-			fmt.Fprint(fw, defineUint32("ETH_HLEN", mac.EthHdrLen))
+			fmt.Fprint(fw, defineUint16("ETH_HLEN", mac.EthHdrLen))
 		}
 	} else {
 		// We want to ensure that the template BPF program always has "LXC_IP"

